### PR TITLE
Routines to generate output product names based on pipeline-supplied visit information

### PR DESCRIPTION
**Pull request by mdlpstsci**
This branch is effectively a fresh start to PR#322.  A single new file has been added to the updated master code to support this functionality.

### CODE REVIEW

File: drizzlepac/generate_final_product_filenames.py [hunk 1]
```diff
@@ -1,15 +1,172 @@
...
+                                                                'ia1s70iwq_flt.fits']}
+
+    """
+    colnames = ['filename', 'proposal_id', 'program_id', 'obset_id', 'visit_id',
+                'exptime', 'filters', 'detector', 'pathname']
+    obset_table = Table.read(results, format='ascii.fast_no_header', names=colnames)
+    # Add INSTRUMENT column
+    instr = INSTRUMENT_DICT[obset_table['filename'][0][0]]
+    # convert input to an Astropy Table for parsing
+    obset_table.add_column(Column([instr]*len(obset_table)), name='instrument')
```
Comment (stsci-hack) [inline drizzlepac/generate_final_product_filenames.py]:
```suggestion
    obset_table.add_column(Column([instr] * len(obset_table)), name='instrument')
```
PEP8 compliance

File: drizzlepac/generate_final_product_filenames.py [hunk 2]
```diff
@@ -1,15 +1,172 @@
...
+        if det not in obset_tree:
+            obset_tree[det] = {}
+            obset_tree[det][filt] = [(row_info, filename)]
+        else:
+            det_node = obset_tree[det]
+            if filt not in det_node:
+                det_node[filt] = [(row_info, filename)]
+            else:
+                filt_node = det_node[filt]
+                filt_node.append((row_info, filename))
```
Comment (stsci-hack) [inline drizzlepac/generate_final_product_filenames.py]:
```suggestion
                det_node[filt].append((row_info, filename))
```

File: drizzlepac/generate_final_product_filenames.py [hunk 3]
```diff
@@ -1,15 +1,172 @@
...
+        # Initial population of the obset tree for this detector
+        if det not in obset_tree:
+            obset_tree[det] = {}
+            obset_tree[det][filt] = [(row_info, filename)]
+        else:
+            det_node = obset_tree[det]
+            if filt not in det_node:
+                det_node[filt] = [(row_info, filename)]
+            else:
+                filt_node = det_node[filt]
```
Comment (stsci-hack) [inline drizzlepac/generate_final_product_filenames.py]: This line can be removed if suggestion for next line is implemented.  This variable never gets used anywhere else, so no need in creating it at all.

File: drizzlepac/generate_final_product_filenames.py [hunk 4]
```diff
@@ -1,15 +1,172 @@
...
+            # Populate single exposure entry now as well
+            for filename in filter_files:
+                # Parse the first filename[1] to determine if the products are flt or flc
+                if det_indx != prev_det_indx:
+                    filetype = "drc"
+                    if filename[1][10:13].lower().endswith("flt"):
+                        filetype = "drz"
+                    prev_det_indx = det_indx
+                sep = SEP_STR.format(sep_indx)  # keep 80 char wide code
+                sep_info = (filename[0]+" "+filetype).lower()
```
Comment (stsci-hack) [inline drizzlepac/generate_final_product_filenames.py]:
```suggestion
                sep_info = (filename[0] + " " + filetype).lower()
```
PEP8 compliance

File: drizzlepac/generate_final_product_filenames.py [hunk 5]
```diff
@@ -1,15 +1,172 @@
...
+                obset_products[sep] = {'info': sep_info,
+                                       'files': [filename[1]]}
+                # Initialize `info` key for this filter product
+                if not obset_products[fprod]['info']:
+                    # Use all but last entry in filter level info
+                    fp_info = (" ".join(filename[0].split()[:])+" "+filetype).lower()
```
Comment (stsci-hack) [inline drizzlepac/generate_final_product_filenames.py]:
```suggestion
                    fp_info = (" ".join(filename[0].split()[:]) + " " + filetype).lower()
```
PEP8 compliance

File: drizzlepac/generate_final_product_filenames.py [hunk 6]
```diff
@@ -1,15 +1,172 @@
...
+                    obset_products[fprod]['info'] = fp_info
+                # Populate filter product with input filename
+                obset_products[fprod]['files'].append(filename[1])
+                # Initialize `info` key for total detection product
+                if not obset_products[tdp]['info']:
+                    tdp_info = (" ".join(filename[0].split()[:-2])+" "+filename[0].split()[-1]+" "+filetype).lower()
```
Comment (stsci-hack) [inline drizzlepac/generate_final_product_filenames.py]:
```suggestion
                    tdp_info = (" ".join(filename[0].split()[:-2]) + " " + filename[0].split()[-1] + " " + filetype).lower()
```
PEP8 compliance

File: drizzlepac/generate_final_product_filenames.py [hunk 7]
```diff
@@ -30,177 +187,156 @@ def run_generator(product_category,obs_info):
...
     # parse out obs_info into a list
     obs_info = obs_info.split(" ")
 
     # pad 4-character proposal_id values with leading 0s so that proposal_id is
-    # a 5-character sting.
-    if key != "multivisit mosaic product": # pad
-        obs_info[0] = "{}{}".format("0"*(5-len(obs_info[0])),obs_info[0])
+    # a 5-character string.
+    if category_key != "multivisit mosaic product":  # pad
+        obs_info[0] = "{}{}".format("0"*(5-len(obs_info[0])), obs_info[0])
```
Comment (stsci-hack) [inline drizzlepac/generate_final_product_filenames.py]:
```suggestion
        obs_info[0] = "{}{}".format("0" * (5 - len(obs_info[0])), obs_info[0])
```
PEP8 compliance

File: drizzlepac/generate_final_product_filenames.py [hunk 8]
```diff
@@ -30,177 +187,156 @@ def run_generator(product_category,obs_info):
...
-    ipppssoot = obs_info[5]
+    obset_id    = obs_info[1]
+    instrument  = obs_info[2]
+    detector    = obs_info[3]
+    filter      = obs_info[4]
+    ipppssoot   = obs_info[5]
+    filetype    = obs_info[6]
+    """
 
+    basename = 'hst_'+'_'.join(map(str, obs_info[:5]))+"_"+obs_info[5][:8]+"_"+obs_info[6]
```
Comment (stsci-hack) [inline drizzlepac/generate_final_product_filenames.py]:
```suggestion
    basename = 'hst_' + '_'.join(map(str, obs_info[:5])) + "_" + obs_info[5][:8] + "_" + obs_info[6]
```
PEP8 compliance

File: drizzlepac/generate_final_product_filenames.py [hunk 9]
```diff
@@ -30,177 +187,156 @@ def run_generator(product_category,obs_info):
...
-    filter = obs_info[4]
+    obset_id    = obs_info[1]
+    instrument  = obs_info[2]
+    detector    = obs_info[3]
+    filter      = obs_info[4]
+    ipppssoot   = obs_info[5]
+    filetype    = obs_info[6]
+    """
 
+    basename = 'hst_'+'_'.join(map(str, obs_info[:5]))+"_"+obs_info[5][:6]
```
Comment (stsci-hack) [inline drizzlepac/generate_final_product_filenames.py]:
```suggestion
    basename = 'hst_' + '_'.join(map(str, obs_info[:5])) + "_" + obs_info[5][:6]
```

File: drizzlepac/generate_final_product_filenames.py [hunk 10]
```diff
@@ -30,177 +187,156 @@ def run_generator(product_category,obs_info):
...
+    detector    = obs_info[2]
+    filter      = obs_info[3]
+    filetype    = obs_info[4]
     """
-    group_num = obs_info[0]
-    instrument = obs_info[1]
-    detector = obs_info[2]
-    filter = obs_info[3]
 
+    basename = 'hst_mos'+'_'.join(map(str, obs_info[:4]))
```
Comment (stsci-hack) [inline drizzlepac/generate_final_product_filenames.py]:
```suggestion
    basename = 'hst_mos' + '_'.join(map(str, obs_info[:4]))
```

File: drizzlepac/generate_final_product_filenames.py [hunk 11]
```diff
@@ -30,177 +187,156 @@ def run_generator(product_category,obs_info):
                                   'multivisit mosaic product': multivisit_mosaic_product_filename_generator}
 
     # Determine which name generator to use based on input product_category
-    for key in category_generator_mapping.keys():
-        if product_category.startswith(key):
-            generator_name = category_generator_mapping[key]
-            category_num = product_category.replace(key+" ","")
+    category_key = ""
+    for ikey in category_generator_mapping.keys():
```
Comment (stsci-hack) [inline drizzlepac/generate_final_product_filenames.py]:
```suggestion
    for ikey in category_generator_mapping:
```
Default behaviour for indexing a dict anyway...

**Comment by mdlpstsci** [comment]
Updated the code per review.

File: drizzlepac/generate_final_product_filenames.py [hunk 12]
```diff
@@ -1,15 +1,171 @@
...
+            if filt not in det_node:
+                det_node[filt] = [(row_info, filename)]
+            else:
+                det_node[filt].append((row_info, filename))
+
+    return obset_tree
+
+def create_row_info(row):
+    """Build info string for a row from the obset table"""
+    info_list = [str(row['proposal_id']), row['program_id'], row['instrument'],
```
Comment (stsci-hack) [inline drizzlepac/generate_final_product_filenames.py]:
```suggestion
    info_list = [str(row['proposal_id']), row['obset_id'], row['instrument'],
```

File: drizzlepac/generate_final_product_filenames.py [hunk 13]
```diff
@@ -1,15 +1,171 @@
...
+        ia1s70iwq_flt.fits,11150,A1S,70,70,0.91161000000000003,F160W,IR,/ifs/archive/ops/hst/public/ia1s/ia1s70iwq/ia1s70iwq_flt.fits
+        which are
+        filename, proposal_id, program_id, obset_id, visit_id, exptime, filters, detector, pathname
+
+    Output dict will have format (as needed by further code for creating the
+        product filenames) of:
+
+        obs_info_dict["single exposure product 00"] = {'info': '11150 A1S WFC3 IR F110W IA1S70JTQ',
+                                                        'files':['ia1s70jtq_flt.fits']}
+        obs_info_dict["single exposure product 01"] = {'info': '11150 A1S WFC3 IR F160W IA1S70JWQ',
```
Comment (stsci-hack) [inline drizzlepac/generate_final_product_filenames.py]:
```suggestion
        obs_info_dict["single exposure product 01"] = {'info': '11150 70 WFC3 IR F160W IA1S70JWQ',
```

File: drizzlepac/generate_final_product_filenames.py [hunk 14]
```diff
@@ -1,15 +1,171 @@
...
+    Input will have format of:
+        ia1s70jtq_flt.fits,11150,A1S,70,70,149.232269,F110W,IR,/ifs/archive/ops/hst/public/ia1s/ia1s70jtq/ia1s70jtq_flt.fits
+        ia1s70iwq_flt.fits,11150,A1S,70,70,0.91161000000000003,F160W,IR,/ifs/archive/ops/hst/public/ia1s/ia1s70iwq/ia1s70iwq_flt.fits
+        which are
+        filename, proposal_id, program_id, obset_id, visit_id, exptime, filters, detector, pathname
+
+    Output dict will have format (as needed by further code for creating the
+        product filenames) of:
+
+        obs_info_dict["single exposure product 00"] = {'info': '11150 A1S WFC3 IR F110W IA1S70JTQ',
```
Comment (stsci-hack) [inline drizzlepac/generate_final_product_filenames.py]:
```suggestion
        obs_info_dict["single exposure product 00"] = {'info': '11150 70 WFC3 IR F110W IA1S70JTQ',
```

File: drizzlepac/generate_final_product_filenames.py [hunk 15]
```diff
@@ -1,15 +1,171 @@
...
+                                                        'files':['ia1s70jtq_flt.fits']}
+        obs_info_dict["single exposure product 01"] = {'info': '11150 A1S WFC3 IR F160W IA1S70JWQ',
+                                                        'files':['ia1s70jwq_flt.fits']}
+
+        obs_info_dict["filter product 00"] = {"info": '11150 A1S WFC3 IR F110W',
```
Comment (stsci-hack) [inline drizzlepac/generate_final_product_filenames.py]:
```suggestion
        obs_info_dict["filter product 00"] = {"info": '11150 70 WFC3 IR F110W',
```

File: drizzlepac/generate_final_product_filenames.py [hunk 16]
```diff
@@ -1,15 +1,171 @@
...
+                                              "files":['ia1s70jtq_flt.fits']}
+        obs_info_dict["filter product 01"] = {"info": '11150 A1S WFC3 IR F160W',
```
Comment (stsci-hack) [inline drizzlepac/generate_final_product_filenames.py]:
```suggestion
        obs_info_dict["filter product 01"] = {"info": '11150 70 WFC3 IR F160W',
```

File: drizzlepac/generate_final_product_filenames.py [hunk 17]
```diff
@@ -1,15 +1,171 @@
...
+    # a dict of 'info' and 'files' information
+
+    # Start interpreting the obset table
+    obset_tree = {}
+    for row in obset_table:
+        # Get some basic information from the first row - no need to check
+        # for multiple instruments as data from different instruments will
+        # not be combined.
+        det = row['detector']
+        filt = row['filters']
```
Comment (stsci-hack) [inline drizzlepac/generate_final_product_filenames.py]: Something needs to be done with the value of `filt` here to remove extraneous 'CLEAR' entries and to convert the ';' to '-' as well, unless this conversion is expected to take place later in the code when the final product names are generated.

**Comment by mdlpstsci** [comment]
This PR is being superseded by PR #391.

**Pull request closed without merging**